src/webcrawler/workers/ParseLinksWorker.java

```diff
@@ -0,0 +1,44 @@
+package webcrawler.workers;
+
+import java.util.List;
+import java.util.Vector;
+import javax.swing.SwingWorker;
+import javax.swing.table.DefaultTableModel;
+import webcrawler.crawl.Html;
+import webcrawler.crawl.HtmlPageParser;
+import webcrawler.source.PageSourceReader;
+
+public class ParseLinksWorker extends SwingWorker<Void, Void> {
+
+  private final Html rootPage;
+  private final DefaultTableModel linksTable;
+
+  private final PageSourceReader pageReader;
+  private final HtmlPageParser pageParser;
+
+  public ParseLinksWorker(Html rootPage, DefaultTableModel linksTable,
+      PageSourceReader pageReader, HtmlPageParser pageParser) {
+    this.rootPage = rootPage;
+    this.linksTable = linksTable;
+    this.pageReader = pageReader;
+    this.pageParser = pageParser;
+  }
+
+
+  @Override
+  protected Void doInBackground() throws Exception {
+    Vector<Vector<String>> allData = new Vector<>();
+    for (String link : rootPage.getLinks()) {
+      if (pageReader.isHtml(link)) {
+        Vector<String> res = new Vector<>();
+        final String linkText = pageReader.readPageSource(link);
+        Html htmlLink = pageParser.parse(linkText, link);
+        res.add(link);
+        res.add(htmlLink.getTitle());
+        allData.add(res);
+      }
+    }
+    linksTable.setDataVector(allData, new Vector<>(List.of("URL", "Title")));
+    return null;
```

Comment:
Maybe it is better to avoid returning nulls?

Reply:
Thanks you are right. This worker initially designed to return from doInBackground()